---
hide:
  - navigation
  - toc
---

# S-LDSC Analysis of RVEF

I used [Stratified Linkage Disequilibrium Score Regression](../../Bioinformatics_Concepts/S_LDSC_For_Cell_And_Tissue_ID.md) (S-LDSC) to analyze summary statistics from a GWAS of MRI-determined right ventricular ejection fraction by Pirruccello et al.  

I used the standard reference data sources to associate chromatin regions with cell or tissue types. There reference data sources were:

- [The GTEx Project](../../Data_Sources/GTEx_Project/GTEx_RNAseq_Data.md)
- The Franke lab dataset
- [The Roadmap Epigenetic Project](../../Data_Sources/Roadmap_Epigenetic_Project/Roadmap.md)
- The [Corces et al. ATAC-seq dataset of 13 blood cell types](../../Data_Sources/Corces_Et_Al_Blood_Cell_Epigenetics/Corces_et_al.md).
- The [ImmGen](../../Data_Sources/Immgen_Project/Immgen_Project.md) Project
- The Cahoy Mouse Central Nervous System Dataset

## Results

### GTEx and Franke lab tissue expression data

The plot below illustrates the pattern of S-LDSC p-values across cell types. The table shows the cell types with the lowest p-values.


![gene-expression-sldsc-rvef](https://github.com/user-attachments/assets/1465eb75-246a-484f-9af5-82d512e49884)


| Name                                                           |   Coefficient |   Coefficient_P_value | Reject Null   |
|:---------------------------------------------------------------|--------------:|----------------------:|:--------------|
| Heart_Left_Ventricle                                           |   2.59059e-08 |           4.37654e-05 | True          |
| Muscle_Skeletal                                                |   2.28636e-08 |           0.000112161 | False         |
| A07.541.560.Heart.Ventricles                                   |   2.25644e-08 |           0.000183698 | False         |
| A07.541.Heart                                                  |   2.16562e-08 |           0.000203794 | False         |
| Heart_Atrial_Appendage                                         |   1.91676e-08 |           0.000272105 | False         |
| A07.541.358.Heart.Atria                                        |   1.68989e-08 |           0.00406107  | False         |
| A02.633.567.850.Quadriceps.Muscle                              |   1.82694e-08 |           0.00554677  | False         |
| A07.541.358.100.Atrial.Appendage                               |   1.5286e-08  |           0.00909596  | False         |


Given that our phenotype is related to the contractive power of a ventricle, this makes sense.  I assume that the match with  left ventricular tissue type is caused by transcriptomic overlap between the left and right ventricles

It is also interesting to observe that all the top cell types are musculoskeletal. This indicates that the key expression patterns that LDSC is picking up is likely characteristic of muscle tissue.



### Roadmap Chromatin data

I next applied S-LDSC to the RVEF GWAS using reference data generated by Finucate et al. from the [Roadmap Epigenetics Project](../../Data_Sources/Roadmap_Epigenetic_Project/Roadmap.md)

The results are shown in the graph and table below

![lvef_ldsc_chromatin_plot](https://github.com/user-attachments/assets/67e9497b-3110-4cf0-90c4-c21bf78c4ecb)

| Name                                                                     |   Coefficient |   Coefficient_P_value | Reject Null   |
|:-------------------------------------------------------------------------|--------------:|----------------------:|:--------------|
| Right_Ventricle__H3K27ac                                                 |   3.94707e-07 |           1.33824e-06 | True          |
| Left_Ventricle__H3K27ac                                                  |   2.80259e-07 |           1.34429e-06 | True          |
| Fetal_Heart__H3K9ac                                                      |   3.62558e-07 |           2.31896e-06 | True          |
| Stomach_Smooth_Muscle__H3K4me3                                           |   6.11652e-07 |           2.7594e-06  | True          |
| Fetal_Heart__DNase                                                       |   3.67744e-07 |           5.20833e-06 | True          |
| Right_Atrium__H3K27ac                                                    |   3.42679e-07 |           1.7163e-05  | True          |
| Colon_Smooth_Muscle__H3K4me3                                             |   6.20742e-07 |           7.11116e-05 | True          |
| Skeletal_Muscle_Male__H3K4me3                                            |   5.99152e-07 |           0.000103465 | True          |
| Ovary__H3K4me1                                                           |   2.59535e-07 |           0.000340226 | False         |
| Fetal_Heart__H3K4me1                                                     |   1.3078e-07  |           0.000356224 | False         |
| Ovary__H3K27ac                                                           |   3.01364e-07 |           0.000510812 | False         |
| Colon_Smooth_Muscle__H3K27ac                                             |   1.74983e-07 |           0.000570302 | False         |
| Esoph-GJ_ENTEX__H3K4me3                                                  |   5.49898e-07 |           0.00060134  | False         |
| Heart-Atrial_ENTEX__H3K27ac                                              |   1.24243e-07 |           0.000604863 | False         |
| Colon_Smooth_Muscle__H3K4me1                                             |   1.71125e-07 |           0.000623398 | False         |
| Fetal_Muscle_Leg__H3K4me3                                                |   9.45169e-07 |           0.00067477  | False         |
| Skeletal_Muscle_Female__H3K4me3                                          |   4.72043e-07 |           0.00123306  | False         |
| NHDF-Ad_Adult_Dermal_Fibroblast_Primary_Cells__H3K36me3                  |   4.26679e-07 |           0.00133146  | False         |

Impressively, the most significant tissue here is the right ventricle, consistent with the phenotype.  Moreover, all other significant tissues are types of muscle, suggesting that this LDSC run has also picked up on muscle-specific gene expression patterns.

The presence of "Ovary" at a low but insignificant p value may point to a sex-hormone effect on heart function.


### Imgen Data

No immgen cell types were significant, consistent with RVEF being a mainly non-immune phenotype.

Here are the lowest p values:
| Name                            |   Coefficient |   Coefficient_P_value | Reject Null   |
|:--------------------------------|--------------:|----------------------:|:--------------|
| Tgd.vg4+24ahi.e17.Th            |   2.82104e-08 |            0.00539746 | False         |
| Tgd.vg2+24ahi.e17.Th.v2         |   2.13139e-08 |            0.00648692 | False         |
| CD4.24h.LN                      |   2.13795e-08 |            0.010458   | False         |
| B.Fo.LN                         |   1.39012e-08 |            0.0130979  | False         |
| B1a.PC                          |   1.20947e-08 |            0.0233794  | False         |
| CD8.24h.LN                      |   1.64771e-08 |            0.0254857  | False         |
| DC.pDC.8+.Sp                    |   1.2325e-08  |            0.0278243  | False         |
| LEC.SLN.CFA.d6.v2               |   1.25117e-08 |            0.0288681  | False         |
| preB.FrD.BM                     |   1.15162e-08 |            0.0325545  | False         |
| B.Fo.MLN                        |   1.17467e-08 |            0.0338449  | False         |
| preB.FrD.FL                     |   1.72978e-08 |            0.035241   | False         |
| Tgd.vg2+24ahi.Th                |   1.4413e-08  |            0.038868   | False         |


### Corces et al. ATAC-seq data

Similar to the above, no cell types were significant when we used the Corces ATAC-seq dataset. The lowest p values are below

| Name    |   Coefficient |   Coefficient_P_value | Reject Null   |
|:--------|--------------:|----------------------:|:--------------|
| Erythro |   2.35709e-07 |              0.119    | False         |
| Bcell   |   4.14389e-08 |              0.337901 | False         |
| MEP     |  -3.19943e-08 |              0.632292 | False         |
| MPP     |  -4.04893e-08 |              0.694066 | False         |
| HSC     |  -4.9625e-08  |              0.733438 | False         |
| CMP     |  -6.57733e-08 |              0.768703 | False         |
| CLP     |  -1.05289e-07 |              0.789269 | False         |
| GMP     |  -1.24871e-07 |              0.931607 | False         |


### Cahoy and GTEx-Brain data

No cell types were significant in the cahoy or GTEx brain datasets, indicating that the GWAS did not detect signifiant associations between neurological cell types and right ventricular function.

| Name            |   Coefficient |   Coefficient_P_value | Reject Null   |
|:----------------|--------------:|----------------------:|:--------------|
| Neuron          |   8.00207e-09 |              0.124024 | False         |
| Oligodendrocyte |  -6.04124e-09 |              0.823426 | False         |
| Astrocyte       |  -5.98411e-09 |              0.867185 | False         |⏎

| Name                                    |   Coefficient |   Coefficient_P_value | Reject Null   |
|:----------------------------------------|--------------:|----------------------:|:--------------|
| Brain_Cerebellar_Hemisphere             |   1.10584e-08 |              0.10896  | False         |
| Brain_Cerebellum                        |   7.64012e-09 |              0.196354 | False         |
| Brain_Frontal_Cortex_(BA9)              |   8.0182e-10  |              0.434653 | False         |
| Brain_Cortex                            |   2.58682e-10 |              0.48244  | False         |
| Brain_Substantia_nigra                  |   3.19984e-11 |              0.497729 | False         |
| Brain_Hippocampus                       |  -3.92956e-10 |              0.520903 | False         |
| Brain_Anterior_cingulate_cortex_(BA24)  |  -4.1335e-10  |              0.534662 | False         |
| Brain_Hypothalamus                      |  -1.52283e-09 |              0.613854 | False         |
| Brain_Amygdala                          |  -1.59616e-09 |              0.620004 | False         |
| Brain_Spinal_cord_(cervical_c-1)        |  -4.26574e-09 |              0.754433 | False         |
| Brain_Caudate_(basal_ganglia)           |  -6.2817e-09  |              0.92293  | False         |
| Brain_Putamen_(basal_ganglia)           |  -1.20152e-08 |              0.986481 | False         |
| Brain_Nucleus_accumbens_(basal_ganglia) |  -1.4452e-08  |              0.992423 | False         |⏎     

### Reproducing

To reproduce this analysis, use the [right heart analysis script][mecfs_bio.analysis.right_heart_analysis]