---
hide:
- navigation
- toc
---
# S-LDSC Asthma Analysis



I applied [Stratified Linkage Disequilibrium Score Regression](../../Bioinformatics_Concepts/S_LDSC_For_Cell_And_Tissue_ID.md) (S-LDSC) to summary statistics from Han et al.'s meta-GWAS of asthma[@han2020genome].



## Reference Data Sources

I used the standard reference datasets recommended and preprocessed by the authors of the S-LDSC method.

- [The GTEx Project](../../Data_Sources/GTEx_RNAseq_Data.md)
- The Franke lab dataset
- [The Roadmap Epigenetic Project](../../Data_Sources/Roadmap.md)
- The [Corces et al. ATAC-seq dataset of 13 blood cell types](../../Data_Sources/Corces_et_al.md).
- The [ImmGen](../../Data_Sources/Immgen_Project.md) Project
- The Cahoy Mouse Central Nervous System Dataset

## Results

### GTEx and Franke lab tissue expression data

The results are plotted below

![s-ldsc-asthma-expression](https://github.com/user-attachments/assets/464dc75c-e693-4659-9a94-e42cd58fde11)


The table below shows the tissue/cell types with the lowest p-values.


| Name                                                           |   Coefficient |   Coefficient_P_value | Reject Null   |
|:---------------------------------------------------------------|--------------:|----------------------:|:--------------|
| A11.118.637.555.567.569.T.Lymphocytes                          |   1.7136e-08  |           1.79629e-06 | True          |
| A15.145.229.637.555.567.569.200.CD4.Positive.T.Lymphocytes     |   1.09652e-08 |           3.90918e-05 | True          |
| A15.382.520.604.700.Spleen                                     |   1.19426e-08 |           0.00010796  | True          |
| A02.835.583.443.800.800.Synovial.Fluid                         |   1.28336e-08 |           0.000122159 | True          |
| A15.145.229.Blood.Cells                                        |   1.17864e-08 |           0.000122277 | True          |
| A15.382.490.555.567.Lymphocytes                                |   1.21767e-08 |           0.000125171 | True          |
| A11.118.637.555.567.569.200.700.T.Lymphocytes..Regulatory      |   1.01439e-08 |           0.000130168 | True          |
| A15.382.490.555.567.537.Killer.Cells..Natural                  |   1.1958e-08  |           0.000343613 | True          |
| A15.145.Blood                                                  |   1.10734e-08 |           0.000367411 | True          |
| Spleen                                                         |   7.89811e-09 |           0.000682061 | False         |
| A10.549.Lymphoid.Tissue                                        |   1.19797e-08 |           0.000759795 | False         |
| A15.145.229.637.555.Leukocytes..Mononuclear                    |   1.02734e-08 |           0.000826575 | False         |
| A15.382.490.555.567.622.Lymphocytes..Null                      |   9.92795e-09 |           0.000963336 | False         |
| A11.118.637.Leukocytes                                         |   8.50474e-09 |           0.00106547  | False         |
| A10.549.400.Lymph.Nodes                                        |   9.3136e-09  |           0.00116527  | False         |
| A03.556.249.124.Ileum                                          |   1.06525e-08 |           0.00264645  | False         |
| A03.556.124.684.Intestine..Small                               |   9.03662e-09 |           0.00267725  | False         |
| A15.378.316.580.Monocytes                                      |   6.43006e-09 |           0.00493671  | False         |
| A02.165.Cartilage                                              |   5.82729e-09 |           0.0057413   | False         |
| Esophagus_Muscularis                                           |   5.82619e-09 |           0.00759493  | False         |
| Whole_Blood                                                    |   7.87232e-09 |           0.00966815  | False         |


These results are consistent with the well-known immunological nature of asthma



### Roadmap Chromatin data

I next applied S-LDSC to the asthma GWAS using reference data generated by Finucane et al. from the [Roadmap Epigenetics Project](../../Data_Sources/Roadmap.md).

The following graph and table show the results:


![s-lsdc-chromatin-asthma](https://github.com/user-attachments/assets/894290df-2b8c-40c7-9b09-0013fbdac162)


| Name                                                                     |   Coefficient |   Coefficient_P_value | Reject Null   |
|:-------------------------------------------------------------------------|--------------:|----------------------:|:--------------|
| Primary_T_helper_17_cells_PMA-I_stimulated__H3K4me1                      |   2.28473e-07 |           1.25164e-09 | True          |
| Primary_T_regulatory_cells_from_peripheral_blood__H3K4me1                |   1.65076e-07 |           1.77049e-09 | True          |
| Primary_T_helper_memory_cells_from_peripheral_blood_2__H3K4me1           |   1.52561e-07 |           9.40009e-08 | True          |
| Primary_T_helper_memory_cells_from_peripheral_blood_1__H3K4me1           |   1.78205e-07 |           1.5018e-07  | True          |
| Primary_T_helper_cells_from_peripheral_blood__H3K4me1                    |   1.3524e-07  |           3.70906e-07 | True          |
| Primary_T_helper_cells_PMA-I_stimulated__H3K4me1                         |   1.22911e-07 |           4.08701e-07 | True          |
| Primary_T_helper_17_cells_PMA-I_stimulated__H3K27ac                      |   3.57399e-07 |           5.40299e-07 | True          |
| Spleen_ENTEX__H3K27ac                                                    |   5.85499e-08 |           6.29279e-07 | True          |
| Primary_T_cells_effector_memory_enriched_from_peripheral_blood__H3K4me1  |   2.19465e-07 |           9.80666e-07 | True          |
| Primary_T_cells_from_peripheral_blood__H3K4me1                           |   1.10633e-07 |           1.12969e-06 | True          |
| Primary_Natural_Killer_cells_from_peripheral_blood__DNase                |   2.87209e-07 |           1.27231e-06 | True          |
| Primary_T_helper_cells_PMA-I_stimulated__H3K27ac                         |   1.62782e-07 |           1.43172e-06 | True          |
| Primary_T_cells_from_peripheral_blood__DNase                             |   2.63651e-07 |           6.12566e-06 | True          |
| Primary_T_killer_memory_cells_from_peripheral_blood__H3K4me1             |   1.4703e-07  |           7.76036e-06 | True          |
| Primary_T_cells_from_peripheral_blood__H3K27ac                           |   1.85166e-07 |           9.91999e-06 | True          |
| Primary_T_helper_17_cells_PMA-I_stimulated__H3K4me3                      |   5.72705e-07 |           1.51429e-05 | True          |
| Primary_mononuclear_cells_from_peripheral_blood__H3K9ac                  |   2.56674e-07 |           1.84474e-05 | True          |
| Primary_T_helper_naive_cells_from_peripheral_blood_1__H3K4me1            |   1.42012e-07 |           2.66904e-05 | True          |
| Primary_T_cells_from_cord_blood__DNase                                   |   2.76964e-07 |           4.84287e-05 | True          |
| Primary_T_regulatory_cells_from_peripheral_blood__H3K27ac                |   1.41208e-07 |           6.79159e-05 | True          |
| Primary_T_helper_cells_PMA-I_stimulated__H3K4me3                         |   3.0952e-07  |           0.000146238 | True          |
| Primary_T_helper_memory_cells_from_peripheral_blood_1__H3K4me3           |   3.30703e-07 |           0.000190911 | True          |
| Primary_T_helper_memory_cells_from_peripheral_blood_1__H3K27ac           |   2.38795e-07 |           0.000191918 | True          |
| Primary_T_killer_naive_cells_from_peripheral_blood__H3K4me1              |   1.03862e-07 |           0.000234462 | True          |
| Colonic_Mucosa__H3K9ac                                                   |   1.69975e-07 |           0.000258698 | True          |
| Primary_T_helper_17_cells_PMA-I_stimulated__H3K36me3                     |   1.21335e-07 |           0.000361052 | True          |
| Primary_T_regulatory_cells_from_peripheral_blood__H3K4me3                |   2.56492e-07 |           0.000408582 | True          |
| Spleen_ENTEX__H3K36me3                                                   |   2.85012e-08 |           0.000628698 | False         |
| Primary_Natural_Killer_cells_from_peripheral_blood__H3K4me1              |   1.02027e-07 |           0.000703613 | False         |
| Spleen_ENTEX__H3K4me1                                                    |   1.71395e-07 |           0.00092961  | False         |
| Primary_T_cells_effector_memory_enriched_from_peripheral_blood__H3K27ac  |   2.5195e-07  |           0.00113583  | False         |
| Primary_T_helper_memory_cells_from_peripheral_blood_2__H3K4me3           |   2.28469e-07 |           0.00130371  | False         |
| Primary_T_helper_naive_cells_from_peripheral_blood__H3K9ac               |   3.11764e-07 |           0.00133795  | False         |
| Primary_Natural_Killer_cells_from_peripheral_blood__H3K27ac              |   1.54913e-07 |           0.00141558  | False         |
| Primary_T_helper_cells_PMA-I_stimulated__H3K36me3                        |   7.10304e-08 |           0.00173045  | False         



Again the focus on immunological cell and tissue types is consistent with known biology.


The single significant intestinal cell type is of unclear importance.  It may be noise and/or artifact, or it may reflect the presence of immune cells in the gut wall.

It is also interesting to observe that in both the GTEx and Roadmap analysis, T cells were the most significant, also consistent with known biology.


### ImmGen data

I next applied S-LDSC to the asthma GWAS using ImmGen data as a reference.

The results are given in the plot and table below:


![s-lsdc-asthma-immgen](https://github.com/user-attachments/assets/edfb8cbc-422e-4174-b874-14647ec9219f)

| Name                            |   Coefficient |   Coefficient_P_value | Reject Null   |
|:--------------------------------|--------------:|----------------------:|:--------------|
| T.4Mem.Sp                       |   1.75441e-08 |           2.27642e-07 | True          |
| NKT.4+.Lv                       |   1.45005e-08 |           5.31667e-07 | True          |
| NKT.4-.Sp                       |   1.61158e-08 |           1.19597e-06 | True          |
| T.4Mem44h62l.Sp                 |   1.51641e-08 |           1.68921e-06 | True          |
| LN.TR.14w.B6                    |   1.4607e-08  |           3.05104e-06 | True          |
| ABD.TR.14w.B6                   |   1.61976e-08 |           4.36389e-06 | True          |
| NKT.4-.Lv                       |   1.31215e-08 |           6.96657e-06 | True          |
| T.4Mem44h62l.LN                 |   1.37984e-08 |           1.28985e-05 | True          |
| CD4Control                      |   1.36494e-08 |           5.34825e-05 | True          |
| T.4FP3+25+.Sp                   |   1.10745e-08 |           0.00012789  | True          |
| T.8Eff.Sp.OT1.d10.LisOva        |   1.07401e-08 |           0.000175188 | True          |
| Tgd.vg2-.Sp                     |   9.08137e-09 |           0.000236126 | True          |
| T.8Eff.Sp.OT1.d8.VSVOva         |   1.08524e-08 |           0.000238524 | True          |
| NKT.44+NK1.1+.Th                |   1.09896e-08 |           0.00030989  | True          |
| Tgd.vg2-.act.Sp                 |   8.33486e-09 |           0.000323691 | True          |
| Tgd.vg2+.Sp                     |   1.07736e-08 |           0.000392383 | True          |
| NKT.4+.Sp                       |   1.1554e-08  |           0.000465112 | True          |
| NK.b2m-.Sp                      |   8.96422e-09 |           0.000602208 | True          |
| T.4.Pa.BDC                      |   1.00297e-08 |           0.000664139 | False         |
| T.4Mem.LN                       |   9.51501e-09 |           0.00070779  | False         |
| T.8Eff.Sp.OT1.d8.LisOva         |   9.7501e-09  |           0.000785852 | False         |
| T.4SP69+.Th                     |   1.01601e-08 |           0.000789917 | False         |
| Tgd.vg2+24alo.Th                |   1.27289e-08 |           0.000825648 | False         |
| Tgd.vg2-.Sp.TCRbko              |   8.78036e-09 |           0.000952023 | False         |
| Tgd.vg2+24ahi.Th                |   1.09829e-08 |           0.00101376  | False         |
| T.8Mem.Sp.OT1.d45.LisOva        |   1.01302e-08 |           0.00112147  | False         |


Here, all significant cell types are either T cell or natural killer cells.  The significance of natural killer cells may reflect their real role in asthma, or their strong transcriptomic overlap with T cells[@aguilar2020immgen], or both.

### Corces et al. ATAC-seq data

Next, I applied S-LDSC using the Corces blood-cell ATAC-seq data as a reference.

| Name    |   Coefficient |   Coefficient_P_value | Reject Null   |
|:--------|--------------:|----------------------:|:--------------|
| CD4     |   2.0966e-07  |           3.20707e-05 | True          |
| CD8     |   1.78171e-07 |           8.74402e-05 | True          |
| NK      |   1.52311e-07 |           0.0025469   | False         |
| Bcell   |   4.71654e-08 |           0.0924839   | False         |
| HSC     |   2.82408e-08 |           0.137143    | False         |
| MPP     |   2.86087e-08 |           0.137208    | False         |
| LMPP    |   1.76042e-08 |           0.314867    | False         |
| Mono    |   2.26418e-08 |           0.334303    | False         |
| CLP     |   1.45358e-08 |           0.345982    | False         |
| MEP     |   7.94853e-09 |           0.350748    | False         |
| GMP     |   8.7878e-09  |           0.373874    | False         |
| CMP     |   2.81872e-09 |           0.455809    | False         |
| Erythro |  -8.14228e-08 |           0.931757    | False         |⏎

Yet again, we see strong significance of T-cell types.



### Cahoy and GTEx-Brain data

Consistent with a disease that is not primarily neurological, there were not significant cell types in the Cahoy or GTEx-brain datasets.


| Name            |   Coefficient |   Coefficient_P_value | Reject Null   |
|:----------------|--------------:|----------------------:|:--------------|
| Astrocyte       |   2.68341e-09 |              0.134337 | False         |
| Neuron          |  -7.26051e-11 |              0.51659  | False         |
| Oligodendrocyte |  -3.19043e-09 |              0.905963 | False         |


| Name                                    |   Coefficient |   Coefficient_P_value | Reject Null   |
|:----------------------------------------|--------------:|----------------------:|:--------------|
| Brain_Cerebellum                        |   7.49537e-09 |            0.00285985 | False         |
| Brain_Cerebellar_Hemisphere             |   2.88573e-09 |            0.0952969  | False         |
| Brain_Spinal_cord_(cervical_c-1)        |   7.39534e-10 |            0.336336   | False         |
| Brain_Cortex                            |   5.27257e-10 |            0.373407   | False         |
| Brain_Amygdala                          |   1.55599e-10 |            0.469241   | False         |
| Brain_Frontal_Cortex_(BA9)              |  -4.11044e-10 |            0.622353   | False         |
| Brain_Nucleus_accumbens_(basal_ganglia) |  -8.97294e-10 |            0.716098   | False         |
| Brain_Anterior_cingulate_cortex_(BA24)  |  -1.45618e-09 |            0.853186   | False         |
| Brain_Putamen_(basal_ganglia)           |  -2.16576e-09 |            0.885143   | False         |
| Brain_Caudate_(basal_ganglia)           |  -1.98446e-09 |            0.910509   | False         |
| Brain_Hippocampus                       |  -2.43085e-09 |            0.921089   | False         |
| Brain_Hypothalamus                      |  -2.56127e-09 |            0.928584   | False         |
| Brain_Substantia_nigra                  |  -3.06368e-09 |            0.974794   | False         |⏎  
