---
hide:
  - toc
---

# S-LDSC

[Stratified Linkage Disequilibrium Score Regression](../../Bioinformatics_Concepts/S_LDSC_For_Cell_And_Tissue_ID.md) (S-LDSC)[@finucane2018heritability] was applied to summary statistics from Bellenguez et al.'s 2019 Meta-GWAS of Multisite Chronic Pain (MCP)[@johnston2019genome].

## Reference Data Sources

This analysis used standard reference datasets recommended and preprocessed by the authors of the S-LDSC method:

- [The GTEx Project](../../Data_Sources/GTEx_RNAseq_Data.md)
- The Franke lab dataset
- [The Roadmap Epigenetic Project](../../Data_Sources/Roadmap.md)
- The [Corces et al. ATAC-seq dataset of 13 blood cell types](../../Data_Sources/Corces_et_al.md).
- The [ImmGen](../../Data_Sources/Immgen_Project.md) Project
- The Cahoy Mouse Central Nervous System Dataset

## Results

### GTEx and Franke lab tissue expression data

Similarly to the [GTEx MAGMA results](./2_MAGMA-GTEx.md), all of the significant tissues in the S-LDSC analysis were related to the brain. Here are the cell/tissue types which were significant after correction using FDR < 0.01:

![s-ldsc_gene-expression](https://github.com/user-attachments/assets/306bb51c-fcd3-402c-84ff-7b6996067e73)

| Name                                                           |   Coefficient |   Coefficient_P_value | Reject Null   |
|:---------------------------------------------------------------|--------------:|----------------------:|:--------------|
| Brain_Putamen_(basal_ganglia)                                  |   4.34441e-09 |           3.82164e-06 | True          |
| Brain_Hippocampus                                              |   3.82353e-09 |           1.05092e-05 | True          |
| Brain_Anterior_cingulate_cortex_(BA24)                         |   4.01219e-09 |           1.20613e-05 | True          |
| A08.186.211.464.Limbic.System                                  |   4.46162e-09 |           1.69685e-05 | True          |
| A08.186.211.Brain                                              |   4.63755e-09 |           2.05309e-05 | True          |
| A08.186.211.730.885.287.500.Cerebral.Cortex                    |   4.40563e-09 |           2.35024e-05 | True          |
| Brain_Caudate_(basal_ganglia)                                  |   3.86268e-09 |           2.77883e-05 | True          |
| A08.186.211.464.710.225.Entorhinal.Cortex                      |   4.31141e-09 |           3.79579e-05 | True          |
| Brain_Nucleus_accumbens_(basal_ganglia)                        |   3.90441e-09 |           3.98038e-05 | True          |
| Brain_Substantia_nigra                                         |   3.64814e-09 |           7.10008e-05 | True          |
| A08.186.211.464.405.Hippocampus                                |   4.02239e-09 |           9.10287e-05 | True          |
| A08.186.211.132.Brain.Stem                                     |   3.93812e-09 |           9.55746e-05 | True          |
| Brain_Cortex                                                   |   3.33234e-09 |           0.000107207 | True          |
| Brain_Frontal_Cortex_(BA9)                                     |   3.37624e-09 |           0.000121439 | True          |
| Brain_Hypothalamus                                             |   3.49643e-09 |           0.000125905 | True          |
| A08.186.211.730.885.287.500.670.Parietal.Lobe                  |   3.92277e-09 |           0.000169976 | True          |
| A08.186.211.730.885.287.500.571.735.Visual.Cortex              |   3.86084e-09 |           0.000205957 | True          |
| A09.371.729.Retina                                             |   3.80165e-09 |           0.000437482 | True          |
| Brain_Cerebellar_Hemisphere                                    |   3.2938e-09  |           0.000438075 | True          |
| Brain_Cerebellum                                               |   3.3889e-09  |           0.000656342 | True          |



### Roadmap chromatin data

Next, reference data generated by Finucane et al.[@finucane2018heritability] from the [Roadmap Epigenetics Project](../../Data_Sources/Roadmap.md) was used in S-LDSC.

In this analysis, almost all significant cell types were again related to the brain, though there were also significant cell types related to the thymus and pancreas. The results can be seen in the plot and table below.

![s-ldsc_chromatin](https://github.com/user-attachments/assets/8adec6b0-c0c8-494c-96b9-5ac0540d6fb1)

| Name                                                                     |   Coefficient |   Coefficient_P_value | Reject Null   |
|:-------------------------------------------------------------------------|--------------:|----------------------:|:--------------|
| Fetal_Brain_Male__DNase                                                  |   1.43001e-07 |           2.73722e-14 | True          |
| Brain_Dorsolateral_Prefrontal_Cortex__H3K27ac                            |   6.61265e-08 |           2.54719e-12 | True          |
| Fetal_Brain_Female__DNase                                                |   1.47145e-07 |           2.80074e-12 | True          |
| Fetal_Brain_Male__H3K4me1                                                |   5.76271e-08 |           4.75756e-11 | True          |
| Fetal_Brain_Female__H3K4me3                                              |   2.05782e-07 |           8.23918e-11 | True          |
| Brain_Germinal_Matrix__H3K4me3                                           |   2.18197e-07 |           3.26595e-10 | True          |
| Fetal_Brain_Female__H3K4me1                                              |   8.26879e-08 |           1.78154e-09 | True          |
| Brain_Dorsolateral_Prefrontal_Cortex__H3K9ac                             |   1.36486e-07 |           3.0457e-09  | True          |
| Brain_Angular_Gyrus__H3K27ac                                             |   5.53797e-08 |           6.53728e-09 | True          |
| Brain_Inferior_Temporal_Lobe__H3K27ac                                    |   4.16002e-08 |           9.83202e-09 | True          |
| Brain_Cingulate_Gyrus__H3K9ac                                            |   1.05355e-07 |           1.33459e-08 | True          |
| Brain_Anterior_Caudate__H3K27ac                                          |   4.3185e-08  |           3.52913e-08 | True          |
| Brain_Germinal_Matrix__H3K4me1                                           |   9.23593e-08 |           4.07106e-08 | True          |
| Brain_Inferior_Temporal_Lobe__H3K4me3                                    |   1.45077e-07 |           4.87189e-08 | True          |
| Ganglion_Eminence_derived_primary_cultured_neurospheres__H3K4me3         |   1.55603e-07 |           5.02739e-08 | True          |
| Brain_Dorsolateral_Prefrontal_Cortex__H3K4me1                            |   6.95193e-08 |           5.60795e-08 | True          |
| Cortex_derived_primary_cultured_neurospheres__H3K4me3                    |   1.8869e-07  |           3.12638e-07 | True          |
| Brain_Hippocampus_Middle__H3K4me3                                        |   1.02857e-07 |           3.66736e-07 | True          |
| Brain_Angular_Gyrus__H3K9ac                                              |   9.12812e-08 |           9.82823e-07 | True          |
| Brain_Dorsolateral_Prefrontal_Cortex__H3K4me3                            |   1.3679e-07  |           1.31053e-06 | True          |
| Brain_Cingulate_Gyrus__H3K4me3                                           |   1.35205e-07 |           1.49391e-06 | True          |
| Brain_Cingulate_Gyrus__H3K27ac                                           |   3.84211e-08 |           2.11232e-06 | True          |
| Brain_Anterior_Caudate__H3K4me1                                          |   4.59498e-08 |           2.50698e-06 | True          |
| Thymus__H3K36me3                                                         |   6.10717e-08 |           4.4298e-06  | True          |
| Brain_Angular_Gyrus__H3K4me1                                             |   4.81968e-08 |           7.71062e-06 | True          |
| Brain_Angular_Gyrus__H3K4me3                                             |   1.30377e-07 |           1.11768e-05 | True          |
| Brain_Inferior_Temporal_Lobe__H3K9ac                                     |   6.47697e-08 |           1.46867e-05 | True          |
| Brain_Anterior_Caudate__H3K9ac                                           |   7.81552e-08 |           2.02698e-05 | True          |
| Brain_Anterior_Caudate__H3K4me3                                          |   1.0891e-07  |           3.78601e-05 | True          |
| Brain_Hippocampus_Middle__H3K27ac                                        |   2.76599e-08 |           4.79293e-05 | True          |
| Brain_Cingulate_Gyrus__H3K4me1                                           |   3.80237e-08 |           5.54432e-05 | True          |
| Brain_Hippocampus_Middle__H3K4me1                                        |   2.86062e-08 |           7.46838e-05 | True          |
| Brain_Inferior_Temporal_Lobe__H3K4me1                                    |   4.24281e-08 |           0.00014697  | True          |
| Brain_Substantia_Nigra__H3K4me3                                          |   9.29301e-08 |           0.000196596 | True          |
| Brain_Germinal_Matrix__H3K36me3                                          |   4.96089e-08 |           0.000445616 | True          |
| Ganglion_Eminence_derived_primary_cultured_neurospheres__H3K4me1         |   3.96413e-08 |           0.000473806 | True          |
| Pancreatic_Islets__H3K27ac                                               |   1.04942e-07 |           0.00051186  | True          |
| Fetal_Thymus__H3K36me3                                                   |   1.97113e-08 |           0.000667859 | True          |

### ImmGen data

In the ImmGen dataset, there were no significant cell types. The following graph and table show the cell types with the lowest p values.

![s-ldsc_immgen](https://github.com/user-attachments/assets/63e56a7b-1b2d-4e30-9790-693296239930)


| Name                            |   Coefficient |   Coefficient_P_value | Reject Null   |
|:--------------------------------|--------------:|----------------------:|:--------------|
| T.DPsm.Th                       |   2.63206e-09 |             0.0065909 | False         |
| T.8Eff.Sp.OT1.d15.LisOva        |   2.24374e-09 |             0.0173596 | False         |
| Tgd.vg2+24ahi.Th                |   2.24909e-09 |             0.0210461 | False         |
| T.8Eff.Sp.OT1.d8.VSVOva         |   1.94559e-09 |             0.0306731 | False         |
| T.DP69+.Th.v2                   |   2.07881e-09 |             0.0313165 | False         |


### Corces et al. ATAC-seq data

There were no significant hematopoietic-related cell types in the Corces ATAC-seq dataset.

| Name    |   Coefficient |   Coefficient_P_value | Reject Null   |
|:--------|--------------:|----------------------:|:--------------|
| HSC     |   9.54732e-09 |              0.237687 | False         |
| GMP     |   7.81531e-09 |              0.273387 | False         |
| LMPP    |   7.72366e-09 |              0.306756 | False         |
| CMP     |   5.13328e-09 |              0.335868 | False         |
| Erythro |   7.49673e-09 |              0.355135 | False         |
| MEP     |   3.51756e-09 |              0.394091 | False         |
| MPP     |   2.5947e-09  |              0.416588 | False         |
| Bcell   |  -7.28211e-09 |              0.717223 | False         |
| Mono    |  -1.5232e-08  |              0.775996 | False         |
| NK      |  -1.02646e-08 |              0.791595 | False         |
| CLP     |  -1.81887e-08 |              0.872501 | False         |
| CD4     |  -2.39266e-08 |              0.968438 | False         |
| CD8     |  -2.37645e-08 |              0.971765 | False         |


### Cahoy and GTEx-Brain data

There were no significant cell types when analyzing the neurological cell types in the Cahoy dataset using S-LDSC:

| Name            |   Coefficient |   Coefficient_P_value | Reject Null   |
|:----------------|--------------:|----------------------:|:--------------|
| Neuron          |   2.57912e-09 |             0.0117611 | False         |
| Oligodendrocyte |   8.92979e-10 |             0.217456  | False         |
| Astrocyte       |  -8.78228e-10 |             0.816394  | False         |

Similarly, there were no significant tissues when testing the GTEx brain tissues using S-LDSC. This contrasts with [MAGMA testing](./2_MAGMA-GTEx.md) on the same tissues revealing several significant brain tissues.

| Name                                    |   Coefficient |   Coefficient_P_value | Reject Null   |
|:----------------------------------------|--------------:|----------------------:|:--------------|
| Brain_Cortex                            |   2.57871e-09 |            0.00388189 | False         |
| Brain_Anterior_cingulate_cortex_(BA24)  |   2.07682e-09 |            0.0103041  | False         |
| Brain_Cerebellum                        |   2.05731e-09 |            0.0341111  | False         |
| Brain_Cerebellar_Hemisphere             |   1.69167e-09 |            0.0639314  | False         |
| Brain_Frontal_Cortex_(BA9)              |   1.15989e-09 |            0.0914603  | False         |
| Brain_Nucleus_accumbens_(basal_ganglia) |   1.05259e-09 |            0.125348   | False         |
| Brain_Caudate_(basal_ganglia)           |   6.12944e-10 |            0.263903   | False         |
| Brain_Putamen_(basal_ganglia)           |   1.48494e-10 |            0.440862   | False         |
| Brain_Amygdala                          |  -5.54761e-10 |            0.722775   | False         |
| Brain_Hippocampus                       |  -1.10526e-09 |            0.887114   | False         |
| Brain_Hypothalamus                      |  -1.22832e-09 |            0.904589   | False         |
| Brain_Spinal_cord_(cervical_c-1)        |  -1.70661e-09 |            0.97087    | False         |
| Brain_Substantia_nigra                  |  -2.03971e-09 |            0.986552   | False         |
