---
hide:
  - navigation
  - toc
---
# S-LDSC DBP Analysis



I applied [Stratified Linkage Disequilibrium Score Regression](../../Bioinformatics_Concepts/S_LDSC_For_Cell_And_Tissue_ID.md) (S-LDSC) to summary statistics from Keaton et al.'s meta-GWAS of blood pressure[@keaton2024genome].  Here I focused on diastolic blood pressure.



## Reference Data Sources

I used the standard reference datasets recommended and preprocessed by the authors of the S-LDSC method.

- [The GTEx Project](../../Data_Sources/GTEx_Project/GTEx_RNAseq_Data.md)
- The Franke lab dataset
- [The Roadmap Epigenetic Project](../../Data_Sources/Roadmap_Epigenetic_Project/Roadmap.md)
- The [Corces et al. ATAC-seq dataset of 13 blood cell types](../../Data_Sources/Corces_Et_Al_Blood_Cell_Epigenetics/Corces_et_al.md).
- The [ImmGen](../../Data_Sources/Immgen_Project/Immgen_Project.md) Project
- The Cahoy Mouse Central Nervous System Dataset

## Results

### GTEx and Franke lab tissue expression data

Surprisingly, no cell types were significant using this reference dataset under a false discovery rate of 0.01.

The cell types with the lowest p values are given below:

| Name                                                           |   Coefficient |   Coefficient_P_value | Reject Null   |
|:---------------------------------------------------------------|--------------:|----------------------:|:--------------|
| A05.360.319.679.690.Myometrium                                 |   8.9244e-09  |           0.000945108 | False         |
| A05.810.453.Kidney                                             |   8.58043e-09 |           0.00118444  | False         |
| A05.360.319.679.Uterus                                         |   1.0349e-08  |           0.00173672  | False         |
| Heart_Atrial_Appendage                                         |   6.83951e-09 |           0.0030787   | False         |
| A03.556.875.875.Stomach                                        |   1.05254e-08 |           0.00345681  | False         |
| Heart_Left_Ventricle                                           |   7.51812e-09 |           0.00353501  | False         |
| Artery_Tibial                                                  |   8.10254e-09 |           0.00454901  | False         |
| A03.556.124.684.Intestine..Small                               |   7.02051e-09 |           0.0076707   | False         |


The failure to find a significant cell type given the very large size of the Keaton et al. meta-GWAS may indicate a weakness in this particular reference dataset, or in the Keaton et al. GWAS data, or in the S-LDSC method itself.


Looking at the top non-significant cell types, the presence of arterial and heart-related cell types as contributors to blood pressure is unsurprising.  What is more surprising is the presence of reproductive tissue at low p-values: myometrium and uterus.  Potentially, this could reflect an effect of sex hormones on blood pressure.



### Roadmap Chromatin data

I next applied S-LDSC to the diastolic blood pressure GWAS using reference data generated by Finucate et al. from the [Roadmap Epigenetics Project](../../Data_Sources/Roadmap_Epigenetic_Project/Roadmap.md)

The following graph and table show the results:


| Name                                                                     |   Coefficient |   Coefficient_P_value | Reject Null   |
|:-------------------------------------------------------------------------|--------------:|----------------------:|:--------------|
| Right_Atrium__H3K4me1                                                    |   1.35454e-07 |           1.40926e-07 | True          |
| Right_Ventricle__H3K4me3                                                 |   4.75769e-07 |           1.47786e-07 | True          |
| Right_Atrium__H3K27ac                                                    |   2.27715e-07 |           2.29194e-07 | True          |
| Aorta__H3K27ac                                                           |   1.73281e-07 |           1.31198e-06 | True          |
| Left_Ventricle__H3K27ac                                                  |   1.44065e-07 |           1.37252e-06 | True          |
| Fetal_Kidney__DNase                                                      |   2.02354e-07 |           3.77236e-06 | True          |
| Right_Atrium__H3K4me3                                                    |   4.0257e-07  |           4.91503e-06 | True          |
| Left_Ventricle__H3K4me1                                                  |   1.20989e-07 |           4.97588e-06 | True          |
| Fetal_Lung__H3K4me1                                                      |   1.0149e-07  |           6.22595e-06 | True          |
| Lung__H3K27ac                                                            |   1.73477e-07 |           7.23294e-06 | True          |
| Fetal_Adrenal_Gland__H3K4me3                                             |   5.20417e-07 |           9.76305e-06 | True          |
| Adipose_Nuclei__H3K4me1                                                  |   9.18524e-08 |           9.86468e-06 | True          |
| Colonic_Mucosa__H3K4me1                                                  |   1.73559e-07 |           1.00295e-05 | True          |
| Fetal_Stomach__DNase                                                     |   2.56155e-07 |           1.10398e-05 | True          |
| Lung__H3K4me1                                                            |   1.51356e-07 |           1.18822e-05 | True          |
| Right_Ventricle__H3K4me1                                                 |   1.22013e-07 |           1.3042e-05  | True          |
| Fetal_Stomach__H3K27ac                                                   |   1.08069e-07 |           1.57813e-05 | True          |
| Adipose_Nuclei__H3K27ac                                                  |   8.34657e-08 |           3.44702e-05 | True          |
| Aorta__H3K4me1                                                           |   2.50867e-07 |           3.54758e-05 | True          |
| Right_Ventricle__H3K27ac                                                 |   1.70317e-07 |           5.12232e-05 | True          |
| Ovary__H3K4me3                                                           |   3.5071e-07  |           5.38539e-05 | True          |
| Stomach_Smooth_Muscle__H3K4me3                                           |   2.211e-07   |           6.47863e-05 | True          |
| Left_Ventricle__H3K4me3                                                  |   4.85369e-07 |           7.96825e-05 | True          |
| Fetal_Lung__DNase                                                        |   1.62032e-07 |           8.89937e-05 | True          |
| Fetal_Stomach__H3K4me3                                                   |   4.01725e-07 |           9.7844e-05  | True          |
| Fetal_Stomach__H3K4me1                                                   |   7.13326e-08 |           0.00010137  | True          |
| Duodenum_Smooth_Muscle__H3K4me1                                          |   1.41764e-07 |           0.00012871  | True          |
| Fetal_Kidney__H3K9ac                                                     |   4.31301e-07 |           0.000141293 | True          |
| Skeletal_Muscle_Male__H3K4me1                                            |   7.46959e-08 |           0.000153192 | True          |
| Fetal_Heart__H3K9ac                                                      |   1.47374e-07 |           0.000180017 | True          |
| Skeletal_Muscle_Male__H3K9ac                                             |   1.40079e-07 |           0.000193951 | True          |
| Spleen__H3K4me1                                                          |   1.03075e-07 |           0.000282654 | True          |
| Aorta__H3K4me3                                                           |   5.18442e-07 |           0.000283657 | True          |
| Skeletal_Muscle_Male__H3K4me3                                            |   2.48575e-07 |           0.000312901 | True          |
| Fetal_Muscle_Leg__H3K4me3                                                |   4.9079e-07  |           0.000318039 | True          |
| Fetal_Adrenal_Gland__H3K36me3                                            |   9.39596e-08 |           0.0006646   | True          |
| Small_Intestine__H3K27ac                                                 |   1.42644e-07 |           0.000743932 | True          |
| Fetal_Adrenal_Gland__H3K4me1                                             |   1.42379e-07 |           0.00108434  | False         |
| Colonic_Mucosa__H3K4me3                                                  |   2.74248e-07 |           0.00125257  | False         |
| Psoas_Muscle__H3K4me3                                                    |   2.92573e-07 |           0.00125564  | False         |
| Fetal_Muscle_Leg__H3K27ac                                                |   8.85931e-08 |           0.00157292  | False         |
| Fetal_Adrenal_Gland__H3K27ac                                             |   1.40667e-07 |           0.00162927  | False         |
| Foreskin_Fibroblast_Primary_Cells_skin02__H3K4me3                        |   1.35138e-07 |           0.00165603  | False         |
| Ovary__H3K27ac                                                           |   1.13578e-07 |           0.0017744   | False         |
| Skeletal_Muscle_Female__H3K4me3                                          |   1.65886e-07 |           0.00189594  | False         |
| Rectal_Smooth_Muscle__H3K4me3                                            |   3.193e-07   |           0.00209098  | False         |
| Adipose_Nuclei__H3K9ac                                                   |   1.45823e-07 |           0.00211634  | False         |
| Fetal_Stomach__H3K36me3                                                  |   4.77451e-08 |           0.00236169  | False         |

![chromatin-dbp-sldsc](https://github.com/user-attachments/assets/ec64787c-2c56-45cc-b4bf-63cceaef0f66)


The contrast with the [previous section](S-LDSC-DBP_Analysis.md#gtex-and-franke-lab-tissue-expression-data) is striking.  Whereas using gene expression reference data produced no significant cell types, using epigenetic reference data produces a wide range of significant cell types across multiple categories.

The most signifiant cell types are heart or kidney related, which is biologically plausible.



### Immgen data

No cell types were significant from the Immgen data, consistent with diastolic blood pressure not being a primarily immune condition.

| Name                            |   Coefficient |   Coefficient_P_value | Reject Null   |
|:--------------------------------|--------------:|----------------------:|:--------------|
| DN.SLN.CFA.d6.v2                |   9.12756e-09 |           0.000235668 | False         |
| DN.SLN.v2                       |   8.84128e-09 |           0.000593548 | False         |
| BEC.SLN.OT                      |   1.09635e-08 |           0.000885477 | False         |
| BEC.SLN                         |   1.02689e-08 |           0.00217312  | False         |
| B.T1.Sp                         |   7.36352e-09 |           0.00286334  | False         |
| B.Mem.Sp.v2                     |   8.22723e-09 |           0.00413403  | False         |
| BEC.MLN                         |   9.61318e-09 |           0.00437598  | False         |
| B.FrE.BM                        |   7.62703e-09 |           0.00465458  | False         |
| Fi.MTS15+.Th                    |   6.49713e-09 |           0.00470782  | False         |
| St.31-38-44-.SLN                |   7.1822e-09  |           0.00508416  | False         |
| LEC.SLN.CFA.d6.v2               |   6.93733e-09 |           0.00588418  | False         |
| SC.LT34F.BM                     |   9.02238e-09 |           0.00704497  | False         |
| B.Fo.LN                         |   7.54089e-09 |           0.00963479  | False         |
| B.FrE.FL                        |   7.46381e-09 |           0.014395    | False         |
| SC.LTSL.FL                      |   8.5851e-09  |           0.0175241   | False         |
| T.4int8+.Th                     |   4.59416e-09 |           0.0265722   | False         |



### Corces et al. ATAC-seq data

Again, as expected, there were no significant hematopoietic-related cell types in the Corces ATAC-seq dataset

| Name    |   Coefficient |   Coefficient_P_value | Reject Null   |
|:--------|--------------:|----------------------:|:--------------|
| HSC     |   4.74203e-08 |              0.10657  | False         |
| Erythro |   6.07373e-08 |              0.109014 | False         |
| MPP     |   3.33285e-08 |              0.173504 | False         |
| CMP     |   2.92056e-08 |              0.200355 | False         |
| MEP     |   3.31256e-08 |              0.24754  | False         |
| CLP     |  -1.67977e-09 |              0.516239 | False         |
| LMPP    |  -1.89293e-09 |              0.518985 | False         |
| Bcell   |  -6.86252e-09 |              0.579055 | False         |
| GMP     |  -7.54583e-09 |              0.592731 | False         |
| NK      |  -4.80328e-08 |              0.938961 | False         |
| Mono    |  -1.13595e-07 |              0.985252 | False         |
| CD8     |  -1.15941e-07 |              0.996904 | False         |
| CD4     |  -1.27186e-07 |              0.999554 | False         |⏎


### Cahoy and GTEx-Brain data

There were no significant neurological cell types.  One might expect to find a significant neurological cell type, given how blood pressure is tightly controlled by the nervous system.  The failure to find this may reflect a limitation of the present reference dataset, or of the S-LDSC method itself.

| Name            |   Coefficient |   Coefficient_P_value | Reject Null   |
|:----------------|--------------:|----------------------:|:--------------|
| Neuron          |   3.72999e-09 |              0.067606 | False         |
| Oligodendrocyte |  -2.21628e-09 |              0.815887 | False         |
| Astrocyte       |  -3.53281e-09 |              0.933489 | False         |⏎  


| Name                                    |   Coefficient |   Coefficient_P_value | Reject Null   |
|:----------------------------------------|--------------:|----------------------:|:--------------|
| Brain_Cerebellum                        |   2.85144e-09 |              0.13386  | False         |
| Brain_Cerebellar_Hemisphere             |   2.4761e-09  |              0.156651 | False         |
| Brain_Putamen_(basal_ganglia)           |  -2.62188e-11 |              0.504228 | False         |
| Brain_Frontal_Cortex_(BA9)              |  -1.16454e-10 |              0.523032 | False         |
| Brain_Substantia_nigra                  |  -3.23304e-10 |              0.55205  | False         |
| Brain_Amygdala                          |  -4.74749e-10 |              0.594472 | False         |
| Brain_Anterior_cingulate_cortex_(BA24)  |  -1.66269e-09 |              0.809254 | False         |
| Brain_Caudate_(basal_ganglia)           |  -2.25646e-09 |              0.819346 | False         |
| Brain_Nucleus_accumbens_(basal_ganglia) |  -2.07221e-09 |              0.839876 | False         |
| Brain_Hippocampus                       |  -2.18755e-09 |              0.855084 | False         |
| Brain_Cortex                            |  -2.1191e-09  |              0.876525 | False         |
| Brain_Hypothalamus                      |  -2.26791e-09 |              0.904322 | False         |
| Brain_Spinal_cord_(cervical_c-1)        |  -5.35052e-09 |              0.994565 | False         |⏎
