---
hide:
  - navigation
  - toc
---
# Chr20 Locus

In an attempt to narrow down the DecodeME[@genetics2025initial] GWAS-1 signal, I [fine-mapped](../../../Bioinformatics_Concepts/Fine_Mapping.md) the GWAS-1 hit on chromosome 20 using SUSIE[@wang2020simple].


As a linkage disequilibrium reference, I used a UK Biobank LD matrix hosted on [AWS Open Data](https://registry.opendata.aws/ukbb-ld/).  Because this LD reference uses GRCh37 coordinates, I used liftover to translate the DecodeME GWAS-1 summary statistics to GRCh37.

## Results with $L=10$

The results of running SUSIE on the chromosome 20 DecodeME GWAS-1 locus are plotted below:


[//]: # (![decode_me_susie_chr20]&#40;../../../_figs/decode_me_chr20_46000001_49000001_susie_stackplot.png&#41;)


SUSIE reports that it has found two credible sets, one of which is diffuse, and one of which is highly concentrated. In the table below, the "cs" column indexes the credible set.  SUSIE assumes that each credible set contains exactly one variant.  The "PIP" column corresponds to "Posterior Inclusion Probability": the odds that a given variant is causal, conditional on the data.

??? info "Variant List"

    |   CHR |      POS | EA    | NEA    |      alpha |         mu |        PIP | cs   |
    |------:|---------:|:------|:-------|-----------:|-----------:|-----------:|:-----|
    |    20 | 47743125 | A     | C      | 1          |  0.0465926 | 1          | L2   |
    |    20 | 47541600 | A     | G      | 0.00887267 |  0.0573823 | 0.00887267 | L3   |
    |    20 | 47543604 | A     | G      | 0.00506655 |  0.0572169 | 0.00506655 | L3   |
    |    20 | 47549979 | T     | TA     | 0.00503385 |  0.0572149 | 0.00503385 | L3   |
    |    20 | 47554392 | G     | A      | 0.00641882 |  0.0572868 | 0.00641882 | L3   |
    |    20 | 47621372 | A     | G      | 0.0138843  |  0.0575143 | 0.0138843  | L3   |
    |    20 | 47635715 | A     | G      | 0.0065204  |  0.0572914 | 0.0065204  | L3   |
    |    20 | 47645427 | T     | G      | 0.00550717 |  0.0572415 | 0.00550717 | L3   |
    |    20 | 47653364 | CTTTG | C      | 0.00513438 |  0.0572208 | 0.00513438 | L3   |
    |    20 | 47662823 | G     | T      | 0.00413755 |  0.0571569 | 0.00413755 | L3   |
    |    20 | 47663382 | G     | A      | 0.00811718 |  0.0573561 | 0.00811718 | L3   |
    |    20 | 47667056 | T     | C      | 0.00626154 | -0.0572795 | 0.00626154 | L3   |
    |    20 | 47673230 | GTTT  | G      | 0.0043698  |  0.0571731 | 0.0043698  | L3   |
    |    20 | 47675525 | A     | G      | 0.00475277 | -0.0571979 | 0.00475277 | L3   |
    |    20 | 47676281 | A     | G      | 0.0104569  | -0.0574308 | 0.0104569  | L3   |
    |    20 | 47677017 | GA    | G      | 0.00461178 | -0.057189  | 0.00461178 | L3   |
    |    20 | 47678770 | G     | A      | 0.00483536 |  0.057203  | 0.00483536 | L3   |
    |    20 | 47687238 | T     | TG     | 0.0202136  |  0.0576247 | 0.0202136  | L3   |
    |    20 | 47690139 | C     | T      | 0.0070961  |  0.0573164 | 0.0070961  | L3   |
    |    20 | 47690189 | A     | G      | 0.00536022 |  0.0572335 | 0.00536022 | L3   |
    |    20 | 47690591 | G     | T      | 0.00976042 |  0.0574105 | 0.00976042 | L3   |
    |    20 | 47692392 | C     | T      | 0.0119906  |  0.0574711 | 0.0119906  | L3   |
    |    20 | 47692827 | C     | T      | 0.00728924 |  0.0573243 | 0.00728924 | L3   |
    |    20 | 47695391 | A     | G      | 0.0100713  | -0.0574197 | 0.0100713  | L3   |
    |    20 | 47701339 | G     | A      | 0.0126552  | -0.057487  | 0.0126552  | L3   |
    |    20 | 47703810 | A     | C      | 0.0176051  | -0.0575841 | 0.0176051  | L3   |
    |    20 | 47710558 | A     | ACAG   | 0.00741292 | -0.0573293 | 0.00741292 | L3   |
    |    20 | 47713063 | T     | C      | 0.00804325 | -0.0573534 | 0.00804325 | L3   |
    |    20 | 47714214 | C     | T      | 0.00984487 | -0.057413  | 0.00984487 | L3   |
    |    20 | 47714291 | G     | A      | 0.0069839  | -0.0573117 | 0.0069839  | L3   |
    |    20 | 47714370 | C     | T      | 0.00913367 | -0.0573909 | 0.00913367 | L3   |
    |    20 | 47717405 | C     | T      | 0.0232343  | -0.0576656 | 0.0232343  | L3   |
    |    20 | 47718594 | C     | T      | 0.00826049 | -0.0573613 | 0.00826049 | L3   |
    |    20 | 47719810 | A     | G      | 0.00869984 | -0.0573765 | 0.00869984 | L3   |
    |    20 | 47719949 | C     | T      | 0.00857865 | -0.0573724 | 0.00857865 | L3   |
    |    20 | 47720957 | C     | T      | 0.00772363 | -0.0573414 | 0.00772363 | L3   |
    |    20 | 47721385 | A     | G      | 0.014393   | -0.0575248 | 0.014393   | L3   |
    |    20 | 47721723 | CAGG  | C      | 0.00710944 | -0.057317  | 0.00710944 | L3   |
    |    20 | 47721759 | C     | T      | 0.00655401 | -0.0572929 | 0.00655401 | L3   |
    |    20 | 47721937 | T     | TTTG   | 0.0227495  | -0.0576594 | 0.0227495  | L3   |
    |    20 | 47725607 | G     | A      | 0.0134008  | -0.0575038 | 0.0134008  | L3   |
    |    20 | 47725800 | C     | CTG    | 0.00481925 | -0.0572021 | 0.00481925 | L3   |
    |    20 | 47726129 | C     | T      | 0.00823929 | -0.0573605 | 0.00823929 | L3   |
    |    20 | 47727079 | A     | G      | 0.0257006  | -0.0576952 | 0.0257006  | L3   |
    |    20 | 47727341 | G     | A      | 0.0110221  | -0.0574463 | 0.0110221  | L3   |
    |    20 | 47728829 | G     | GCTTA  | 0.0114223  | -0.0574568 | 0.0114223  | L3   |
    |    20 | 47731158 | C     | T      | 0.0312728  | -0.0577527 | 0.0312728  | L3   |
    |    20 | 47731228 | C     | A      | 0.106306   | -0.0581101 | 0.106306   | L3   |
    |    20 | 47737958 | C     | A      | 0.00630948 | -0.0572817 | 0.00630948 | L3   |
    |    20 | 47739193 | A     | G      | 0.00629615 | -0.0572811 | 0.00629615 | L3   |
    |    20 | 47739415 | T     | C      | 0.0197913  | -0.0576185 | 0.0197913  | L3   |
    |    20 | 47739814 | T     | C      | 0.00913847 | -0.0573911 | 0.00913847 | L3   |
    |    20 | 47741556 | T     | C      | 0.0133342  | -0.0575024 | 0.0133342  | L3   |
    |    20 | 47746627 | AATAG | A      | 0.00383531 | -0.0571345 | 0.00383531 | L3   |
    |    20 | 47748375 | A     | G      | 0.00450821 | -0.0571823 | 0.00450821 | L3   |
    |    20 | 47749296 | C     | CCAAAA | 0.00599058 | -0.0572664 | 0.00599058 | L3   |
    |    20 | 47750377 | A     | C      | 0.0196992  | -0.0576171 | 0.0196992  | L3   |
    |    20 | 47751708 | A     | AAAG   | 0.0152026  | -0.0575409 | 0.0152026  | L3   |
    |    20 | 47753019 | T     | C      | 0.00478323 | -0.0571998 | 0.00478323 | L3   |
    |    20 | 47755075 | A     | G      | 0.00421025 | -0.0571621 | 0.00421025 | L3   |
    |    20 | 47755742 | T     | C      | 0.00433697 | -0.0571709 | 0.00433697 | L3   |
    |    20 | 47757963 | T     | C      | 0.00387817 | -0.0571377 | 0.00387817 | L3   |
    |    20 | 47758549 | T     | C      | 0.0104108  | -0.0574295 | 0.0104108  | L3   |
    |    20 | 47758719 | A     | G      | 0.00380131 | -0.0571318 | 0.00380131 | L3   |
    |    20 | 47760000 | C     | T      | 0.00471366 | -0.0571955 | 0.00471366 | L3   |
    |    20 | 47762700 | A     | G      | 0.0133314  | -0.0575023 | 0.0133314  | L3   |
    |    20 | 47763558 | A     | C      | 0.003858   | -0.0571362 | 0.003858   | L3   |
    |    20 | 47766197 | C     | T      | 0.0043448  | -0.0571714 | 0.0043448  | L3   |
    |    20 | 47766929 | T     | C      | 0.00411454 | -0.0571553 | 0.00411454 | L3   |
    |    20 | 47769157 | A     | G      | 0.00365162 | -0.0571199 | 0.00365162 | L3   |
    |    20 | 47769837 | AT    | A      | 0.00738487 | -0.0573282 | 0.00738487 | L3   |
    |    20 | 47770756 | C     | A      | 0.0151921  | -0.0575407 | 0.0151921  | L3   |
    |    20 | 47771274 | A     | G      | 0.00561524 | -0.0572473 | 0.00561524 | L3   |
    |    20 | 47774166 | G     | A      | 0.00595017 | -0.0572644 | 0.00595017 | L3   |
    |    20 | 47774343 | A     | G      | 0.00407655 | -0.0571525 | 0.00407655 | L3   |
    |    20 | 47779340 | C     | T      | 0.0306726  | -0.057747  | 0.0306726  | L3   |
    |    20 | 47782356 | C     | CCT    | 0.00699669 | -0.0573122 | 0.00699669 | L3   |
    |    20 | 47783095 | C     | T      | 0.0083366  | -0.057364  | 0.0083366  | L3   |
    |    20 | 47783770 | TG    | T      | 0.0293129  | -0.0577337 | 0.0293129  | L3   |
    |    20 | 47792245 | T     | C      | 0.0093256  | -0.057397  | 0.0093256  | L3   |
    |    20 | 47792457 | C     | A      | 0.0262475  | -0.0577013 | 0.0262475  | L3   |
    |    20 | 47795784 | C     | CA     | 0.00515276 | -0.0572218 | 0.00515276 | L3   |
    |    20 | 47797653 | A     | G      | 0.00430375 | -0.0571686 | 0.00430375 | L3   |
    |    20 | 47798664 | C     | T      | 0.00529461 | -0.0572299 | 0.00529461 | L3   |
    |    20 | 47808070 | A     | G      | 0.0176384  | -0.0575846 | 0.0176384  | L3   |
    |    20 | 47815944 | C     | T      | 0.0115221  | -0.0574594 | 0.0115221  | L3   |
    |    20 | 47818567 | T     | C      | 0.00379868 | -0.0571316 | 0.00379868 | L3   |
    |    20 | 47819305 | A     | G      | 0.0095954  | -0.0574054 | 0.0095954  | L3   |
    |    20 | 47820626 | T     | G      | 0.00903159 | -0.0573876 | 0.00903159 | L3   |




## Results with $L=10$ and strict threshold

The original paper on fine-mapping with SUSIE using summary statistics by Zou et al.[zou2022fine] notes that even slight mismatches between the population used to generate the GWAS summary statistics and the population from which the LD matrix is derived can cause SUSIE to become unstable or produce false positives.  Zou et al. suggest that filtering variants with a high likelihood ratio in a likelihood ratio test while also having high z score can mitigate this issue, since these variants are most impacted the mismatch.  By default, [consistent with the SUSIE documentation](https://stephenslab.github.io/susieR/reference/kriging_rss.html), my code filters variants that have a likelihood ratio of 2 or greater and a z score of 2 or greater.  To get sense of the effect of the sensitivity of SUSIE results to this filtering, I 
