---
hide:
- navigation
- toc
---
# Chr17 50M-51M


## Methodology

In an attempt to narrow down the DecodeME[@genetics2025initial] GWAS-1 signal, I [fine-mapped](../../../../Bioinformatics_Concepts/Fine_Mapping.md) the GWAS-1 hit on chromosome 17 using SUSIE[@wang2020simple]. The parameters of my SUSIE runs were the same as for the [chromosome 1 locus](a_Chr1_173M_174M_Locus.md).


## Results

Each of the 4 SUSIE runs produced a single credible set with the same 63 variants, as illustrated by the UpSet plot below:


![chr17_upset](https://github.com/user-attachments/assets/79044a67-d441-4841-b68f-e0bc602eefde)


The distribution of probability mass across these variants was slightly different across the runs, as illustrated by the following upsetplot, which is restricted to variants of $PIP>0.01$.
![chr_17_pip001](https://github.com/user-attachments/assets/a8867989-fb8e-452e-8754-24ee46dbe761)


The stackplot below, for the $L=10$ case, illustrates the typical PIP pattern across variants.

![chr17_stackplot](https://github.com/user-attachments/assets/f449a641-b3f5-4671-b7ae-0369e2c01179)

The credible set at this locus is focused upstream of the CA10 gene.  At a glance, this would seem to make a CA10 the best candidate for further investigation.  

The detailed variants results are below:

??? info "Variant List (L=10)"


    | cs   |   CHR |      POS | EA   | NEA   |      alpha |        mu |        PIP |
    |:-----|------:|---------:|:-----|:------|-----------:|----------:|-----------:|
    | L1   |    17 | 50260366 | T    | C     | 0.055044   | 0.023495  | 0.055161   |
    | L1   |    17 | 50282834 | T    | C     | 0.0412133  | 0.0232911 | 0.0413269  |
    | L1   |    17 | 50287448 | A    | C     | 0.03704    | 0.0232154 | 0.0371532  |
    | L1   |    17 | 50279497 | T    | C     | 0.0357889  | 0.023191  | 0.035902   |
    | L1   |    17 | 50276929 | G    | C     | 0.0353752  | 0.0231828 | 0.0354882  |
    | L1   |    17 | 50294376 | T    | C     | 0.0347934  | 0.023171  | 0.0349065  |
    | L1   |    17 | 50289205 | T    | G     | 0.0338718  | 0.0231519 | 0.0339847  |
    | L1   |    17 | 50290356 | G    | A     | 0.0326279  | 0.0231252 | 0.0327407  |
    | L1   |    17 | 50285434 | A    | G     | 0.0322195  | 0.0231162 | 0.0323321  |
    | L1   |    17 | 50295022 | A    | G     | 0.032076   | 0.0231131 | 0.0321887  |
    | L1   |    17 | 50272817 | TA   | T     | 0.0318892  | 0.0231089 | 0.0320022  |
    | L1   |    17 | 50271517 | A    | G     | 0.0315094  | 0.0231003 | 0.031622   |
    | L1   |    17 | 50270268 | T    | C     | 0.0296969  | 0.023058  | 0.0298095  |
    | L1   |    17 | 50294496 | G    | A     | 0.0295546  | 0.0230546 | 0.0296669  |
    | L1   |    17 | 50271048 | A    | G     | 0.029051   | 0.0230423 | 0.0291633  |
    | L1   |    17 | 50271674 | A    | G     | 0.0277654  | 0.0230099 | 0.0278775  |
    | L1   |    17 | 50274393 | A    | G     | 0.0272457  | 0.0229964 | 0.0273578  |
    | L1   |    17 | 50261933 | CA   | C     | 0.0224699  | 0.0228578 | 0.0225895  |
    | L1   |    17 | 50298441 | A    | C     | 0.0164279  | 0.0226309 | 0.0165402  |
    | L1   |    17 | 50285721 | CT   | C     | 0.0154185  | 0.0225847 | 0.0155312  |
    | L1   |    17 | 50297907 | G    | A     | 0.0143061  | 0.02253   | 0.0144173  |
    | L1   |    17 | 50288321 | T    | G     | 0.0133761  | 0.0224807 | 0.0134877  |
    | L1   |    17 | 50294365 | G    | C     | 0.0129315  | 0.0224559 | 0.013043   |
    | L1   |    17 | 50291040 | T    | C     | 0.0113472  | 0.0223598 | 0.0114599  |
    | L1   |    17 | 50293499 | C    | T     | 0.011319   | 0.022358  | 0.0114317  |
    | L1   |    17 | 50290955 | T    | C     | 0.0108259  | 0.0223251 | 0.0109383  |
    | L1   |    17 | 50299063 | G    | A     | 0.0106136  | 0.0223105 | 0.0107243  |
    | L1   |    17 | 50296779 | C    | A     | 0.0100429  | 0.0222696 | 0.010153   |
    | L1   |    17 | 50317099 | G    | T     | 0.00958189 | 0.0222348 | 0.00969233 |
    | L1   |    17 | 50299079 | A    | G     | 0.00905355 | 0.0221928 | 0.00916326 |
    | L1   |    17 | 50316842 | G    | A     | 0.00902542 | 0.0221904 | 0.00913553 |
    | L1   |    17 | 50316120 | G    | C     | 0.00869143 | 0.0221624 | 0.00880133 |
    | L1   |    17 | 50314751 | A    | G     | 0.00863265 | 0.0221574 | 0.00874245 |
    | L1   |    17 | 50314078 | T    | C     | 0.00860537 | 0.022155  | 0.00871515 |
    | L1   |    17 | 50316095 | C    | T     | 0.00857642 | 0.0221525 | 0.00868625 |
    | L1   |    17 | 50315044 | T    | C     | 0.00834915 | 0.0221325 | 0.0084588  |
    | L1   |    17 | 50301060 | G    | A     | 0.00734718 | 0.0220371 | 0.00745657 |
    | L1   |    17 | 50314350 | T    | C     | 0.00724198 | 0.0220263 | 0.00735097 |
    | L1   |    17 | 50303770 | G    | A     | 0.00693928 | 0.0219944 | 0.00704804 |
    | L1   |    17 | 50298413 | C    | T     | 0.00676585 | 0.0219754 | 0.00687585 |
    | L1   |    17 | 50313611 | G    | C     | 0.00673919 | 0.0219724 | 0.00684788 |
    | L1   |    17 | 50309650 | G    | A     | 0.00665405 | 0.0219629 | 0.00676269 |
    | L1   |    17 | 50299568 | CT   | C     | 0.00665184 | 0.0219626 | 0.00676031 |
    | L1   |    17 | 50300195 | G    | C     | 0.00659815 | 0.0219566 | 0.00670669 |
    | L1   |    17 | 50312430 | G    | A     | 0.00653389 | 0.0219492 | 0.00664245 |
    | L1   |    17 | 50299809 | C    | T     | 0.00640457 | 0.0219342 | 0.00651299 |
    | L1   |    17 | 50300935 | A    | G     | 0.00639552 | 0.0219331 | 0.00650393 |
    | L1   |    17 | 50312701 | T    | C     | 0.00636609 | 0.0219297 | 0.00647456 |
    | L1   |    17 | 50301112 | G    | A     | 0.00625822 | 0.0219168 | 0.0063666  |
    | L1   |    17 | 50309626 | T    | G     | 0.00624581 | 0.0219153 | 0.00635419 |
    | L1   |    17 | 50313435 | C    | A     | 0.006139   | 0.0219024 | 0.00624731 |
    | L1   |    17 | 50311547 | G    | C     | 0.00610706 | 0.0218984 | 0.00621535 |
    | L1   |    17 | 50309871 | A    | C     | 0.00604867 | 0.0218912 | 0.00615692 |
    | L1   |    17 | 50300731 | A    | G     | 0.00586534 | 0.021868  | 0.00597344 |
    | L1   |    17 | 50301115 | T    | C     | 0.00566136 | 0.0218413 | 0.00576936 |
    | L1   |    17 | 50308625 | G    | A     | 0.00536093 | 0.0218001 | 0.00546875 |
    | L1   |    17 | 50308600 | G    | A     | 0.0053587  | 0.0217998 | 0.00546652 |
    | L1   |    17 | 50291181 | A    | C     | 0.00402888 | 0.0215831 | 0.00413852 |
    | L1   |    17 | 50302285 | G    | A     | 0.00369619 | 0.0215172 | 0.00380575 |
    | L1   |    17 | 50296459 | T    | G     | 0.00315961 | 0.0213967 | 0.00326817 |
    | L1   |    17 | 50302857 | T    | C     | 0.00315112 | 0.0213946 | 0.00325992 |
    | L1   |    17 | 50303110 | G    | C     | 0.0030557  | 0.0213709 | 0.00316436 |
    | L1   |    17 | 50303521 | T    | C     | 0.0030314  | 0.0213648 | 0.00314003 |‚èé  